diff --git a/arch/arm64/configs/exynos2100-unified_defconfig b/arch/arm64/configs/exynos2100-unified_defconfig
index 1be9993954be..73d523bd2034 100755
--- a/arch/arm64/configs/exynos2100-unified_defconfig
+++ b/arch/arm64/configs/exynos2100-unified_defconfig
@@ -3565,7 +3565,8 @@ CONFIG_VIDEOBUF2_DMA_SG=m
 # CONFIG_MALI_DDK_VALHALL_R26P0 is not set
 # CONFIG_MALI_DDK_VALHALL_R32P1 is not set
 # CONFIG_MALI_DDK_VALHALL_R38P0 is not set
-CONFIG_MALI_DDK_VALHALL_R38P1=y
+# CONFIG_MALI_DDK_VALHALL_R38P1 is not set
+CONFIG_MALI_DDK_VALHALL_R49P0=y
 CONFIG_MALI_MIDGARD=m
 # CONFIG_MALI_GATOR_SUPPORT is not set
 # CONFIG_MALI_MIDGARD_DVFS is not set
@@ -3578,6 +3579,7 @@ CONFIG_MALI_EXPERT=y
 # CONFIG_MALI_NO_MALI is not set
 CONFIG_MALI_REAL_HW=y
 # CONFIG_MALI_SYSTEM_TRACE is not set
+# CONFIG_MALI_TRACE_POWER_GPU_WORK_PERIOD is not set
 # CONFIG_MALI_2MB_ALLOC is not set
 # CONFIG_MALI_PWRSOFT_765 is not set
 # CONFIG_MALI_MEMORY_FULLY_BACKED is not set
diff --git a/drivers/gpu/arm/Kbuild b/drivers/gpu/arm/Kbuild
index ff33eca1b011..ed1a07f44571 100755
--- a/drivers/gpu/arm/Kbuild
+++ b/drivers/gpu/arm/Kbuild
@@ -42,3 +42,6 @@ endif
 ifeq ($(CONFIG_MALI_DDK_VALHALL_R38P1),y)
 	obj-$(CONFIG_MALI_MIDGARD) += v_r38p1/
 endif
+ifeq ($(CONFIG_MALI_DDK_VALHALL_R49P0),y)
+	obj-$(CONFIG_MALI_MIDGARD) += bv_r49p0/
+endif
diff --git a/drivers/gpu/arm/Kconfig b/drivers/gpu/arm/Kconfig
index 312a43a6066a..46257d67491f 100755
--- a/drivers/gpu/arm/Kconfig
+++ b/drivers/gpu/arm/Kconfig
@@ -57,6 +57,9 @@ config MALI_DDK_VALHALL_R38P0
 config MALI_DDK_VALHALL_R38P1
 	bool "Valhall r38p1 driver"
 
+config MALI_DDK_VALHALL_R49P0
+	bool "Valhall r49p0 driver"
+
 endchoice
 
 if MALI_DDK_VALHALL_R20P0
@@ -91,4 +94,8 @@ if MALI_DDK_VALHALL_R38P1
 source "drivers/gpu/arm/v_r38p1/Kconfig"
 endif
 
+if MALI_DDK_VALHALL_R49P0
+source "drivers/gpu/arm/bv_r49p0/Kconfig"
+endif
+
 endmenu
diff --git a/drivers/gpu/arm/bv_r49p0/Kconfig b/drivers/gpu/arm/bv_r49p0/Kconfig
index 0b50131..a94b48e 100644
--- a/drivers/gpu/arm/bv_r49p0/Kconfig
+++ b/drivers/gpu/arm/bv_r49p0/Kconfig
@@ -370,7 +370,7 @@ config MALI_HW_ERRATA_1485982_USE_CLOCK_ALTERNATIVE
 endif
 
 config MALI_ARBITRATION
-	boolean "Enable Virtualization reference code"
+	bool "Enable Virtualization reference code"
 	depends on MALI_MIDGARD
 	default n
 	help
diff --git a/drivers/gpu/arm/bv_r49p0/mmu/mali_kbase_mmu.c b/drivers/gpu/arm/bv_r49p0/mmu/mali_kbase_mmu.c
index 9153b92..7c63164 100644
--- a/drivers/gpu/arm/bv_r49p0/mmu/mali_kbase_mmu.c
+++ b/drivers/gpu/arm/bv_r49p0/mmu/mali_kbase_mmu.c
@@ -303,15 +303,10 @@ static void kbase_mmu_sync_pgd_gpu(struct kbase_device *kbdev, struct kbase_cont
 
 static void kbase_mmu_sync_pgd_cpu(struct kbase_device *kbdev, dma_addr_t handle, size_t size)
 {
-	/* In non-coherent system, ensure the GPU can read
-	 * the pages from memory
+	/* Ensure that the GPU can read page-table updates. This mirrors the
+	 * unconditional synchronization in Google's official r49 integration.
 	 */
-#if IS_ENABLED(CONFIG_MALI_EXYNOS_LLC)
-	if (kbdev->system_coherency != COHERENCY_ACE)
-#else
-	if (kbdev->system_coherency == COHERENCY_NONE)
-#endif
-		dma_sync_single_for_device(kbdev->dev, handle, size, DMA_TO_DEVICE);
+	dma_sync_single_for_device(kbdev->dev, handle, size, DMA_TO_DEVICE);
 }
 
 /**
